### ISSUE TITLE
the size of a 3D Arrow head plotted with Axes3D.quiver

### ISSUE THREAD

**Issue by arthurzqhu** [matplotlib/matplotlib#11746]
### Bug report

**Bug summary**

When plotting 3D arrows, there are no arguments in `Axes3D.quiver()` that can be used to changed the shape/size of the head of a 3D arrow. Nor is there any related attribute in `Line3DCollection` that can be used. 

The size of a 3D arrow head seems to be dependent on the scale of the plot, i.e., `xlim`, `ylim`, and `zlim`. In the example below, a change in the scale of `z` causes the arrow head to "appear" or "disappear".

To make things easier to read, my observation of this issue will be posted in the first comment of this post.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

x = np.zeros(10)
y = np.zeros(10)
z = np.arange(10)*100 # remove *100 and the arrow heads will reappear.
dx = np.zeros(10)
dy = np.arange(10)
dz = np.zeros(10)

fig = plt.figure()
ax = fig.gca(projection='3d')
ax.quiver(x, y, z, dx, dy, dz)
ax.set_ylim(0,10)

plt.show()
```

**Actual outcome**

![norm_before](https://user-images.githubusercontent.com/3490251/43068052-259a92c6-8e1e-11e8-9d0b-38f25bc4e47a.png)

**Expected outcome**

**_Notice the difference in the scale of z axis._**
![norm_after](https://user-images.githubusercontent.com/3490251/43068084-3af6605a-8e1e-11e8-8dfd-5b1b2275dc8a.png)
```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

x = np.zeros(10)
y = np.zeros(10)
z = np.arange(10)
dx = np.zeros(10)
dy = np.arange(10)
dz = np.zeros(10)

fig = plt.figure()
ax = fig.gca(projection='3d')
ax.quiver(x, y, z, dx, dy, dz)
ax.set_ylim(0,10)

plt.show()
```

**Matplotlib version**

  * Operating system: macOS 10.13.6
  * Matplotlib version: 2.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 2.7.15
  * Jupyter version (if applicable): 5.6.0
  * Other libraries: 

matplotlib and python were installed with Anaconda.

**Comment by arthurzqhu** [comment]
![photo jul 23 02 55 58](https://user-images.githubusercontent.com/3490251/43070753-8271c9e0-8e25-11e8-8869-ae57edb350e6.jpg)

First, please allow me to define some terms regarding the arrow itself. As shown in the figure, let's say that θ is the angle between the head and the body of an arrow, b is the distance between the far end of the head and the body, and c is the length of the arrow itself.

After taking a closer look, it seems to me that, when the span of z is really big compared to the span of y, the arrow heads are still there (obviously), but they are just mostly overlapped with the body of the arrow. And the reason that they are overlapped I think is because `b` (from the figure) is determined by the size of the arrow, which here is in y-direction, but its size is reflected on the z axis (since the head and the body of the arrow are lying on a plane parallel to z-axis). For example, if `b = 0.1*c` and `c = 10`, then `b = 1`. But if `z` spans over 900 units, then apparently `b = 1` would be minuscule compared to that and thus θ would be very close to 0, making the head almost completely overlaps with the body. I think the correct way to draw an arrow is to set θ as constant (or preferably configurable) so that the appearance of the arrow can be maintained or arbitrarily altered.

(After all, what makes it an arrow if we don't know where its head is pointing?)

**Comment by arthurzqhu** [comment]
@WeatherGod Thanks for the update. Guess I'll stick with my workaround for a while then. 😅

**Comment by jxmaster** [comment]
I wonder if this thing is different now?

**Comment by refmitchell** [comment]
Using matplotlib 3.1.1. As a workaround you could try using the `arrow_length_ratio` argument (by default `arrow_length_ratio=0.3`). I need to plot some of my own axes relative to the centre of a half-sphere. With the default setting I get this:

![DevAxes](https://user-images.githubusercontent.com/10943932/66206257-7ba74f00-e6a7-11e9-9853-1f1918c10ad3.png)

When I set `arrow_length_ratio=0.1` I get this: 
![DevAxesRatioSet](https://user-images.githubusercontent.com/10943932/66206258-7ba74f00-e6a7-11e9-9e74-91baa87e795a.png)

The argument scales the arrow head with respect to the length of the body, documented [here](https://matplotlib.org/mpl_toolkits/mplot3d/tutorial.html?highlight=3d%20quiver%20plot#quiver). This was the best solution I found for my work, hope this is useful and actually answers the issue raised above.

**Comment by arthurzqhu** [comment]
> Using matplotlib 3.1.1. As a workaround you could try using the `arrow_length_ratio` argument (by default `arrow_length_ratio=0.3`). I need to plot some of my own axes relative to the centre of a half-sphere. With the default setting I get this:
> 
> ![DevAxes](https://user-images.githubusercontent.com/10943932/66206257-7ba74f00-e6a7-11e9-9853-1f1918c10ad3.png)
> 
> When I set `arrow_length_ratio=0.1` I get this:
> ![DevAxesRatioSet](https://user-images.githubusercontent.com/10943932/66206258-7ba74f00-e6a7-11e9-9e74-91baa87e795a.png)
> 
> The argument scales the arrow head with respect to the length of the body, documented [here](https://matplotlib.org/mpl_toolkits/mplot3d/tutorial.html?highlight=3d%20quiver%20plot#quiver). This was the best solution I found for my work, hope this is useful and actually answers the issue raised above.

That's exactly what I did for my workaround as well. I was too lazy to type it up here since the relevant project was over, so thanks for posting this.

**Comment by ZhangYuef** [comment]
Using Matplotlib 3.3.2, setting `arrow_length_ratio=0.3` to other values does not solve my problem.  Please help to work around it.

**Comment by timhoffm** [comment]
Indeed, `arrow_length_ratio` is not a general workaround. The underlying problem here is that the arrow head is drawn in data coordinates. If the z-axis is much larger than the x/y axis, the head gets visually squeezed to the shaft. `arrow_length_ratio` only specifies the length of the head. It does not modify the opening angle. I'm sorry to say, but I don't know of a reasonable workaround.

**Comment by ZhangYuef** [comment]
@timhoffm Thanks for the answering, I understand the source of problem now. I notice this issue is raised in 2018.6 and is still alive until today. Just curious, how difficult is that fixing this problem in 3-d data coordinates representation?

**Comment by timhoffm** [comment]
@ZhangYuef This is more difficult that it might appear at first glance. The quick and dirty solution would be to scale the arrow opening angle depending on the ratios of the rendered axes (e.g. pixel-size of one unit of x vs. one unit of z). But even that does not work because the arrow artist does not have that information; and that information can change e.g. when axes limits or the figure size are changed.
Another simplified approach would be to take the arrow directions from a projection of the 3d data space to the 2d rendering area and then add the heads in the 2d image later. We do not have a way of splitting drawing like that. Also it still has the disadvantage that the arrow heads are flat in the drawing space. That reduces the 3d impression of the arrows and may be  unwanted.
I'm not even sure what the best approach is if we had any machinery we want in place: The arrow head direction and head orientation comes from 3d data space, but the head size is data independent. (Would be even more complicated if you want a depth effect; i.e. arrows further away from the viewer are smaller.) Arrow heads are surprisingly complex in 3d.

**Comment by mocquin** [comment]
Feature wanted  ! especially since the `arrow_length_ratio` has to be adapted depending on the data you want to plot
